### ISSUE TITLE
bad `free -m` output on LX zones over their memcap

### ISSUE THREAD

**Issue by zadunn** [joyent/illumos-joyent#132]
Background:

While investigating a slow compute node, I found a LX zone that's over it's zone memcap.  Logging into the zone I tried `free -m` and got the following:

```
root@mongo-replmem-007:/home/zdunn# free -m
             total       used       free     shared    buffers     cached
Mem:         65536      65604 18014398509481916          0          0          0
-/+ buffers/cache:      65604 18014398509481916
Swap:       131072          0     131072
```

errata from IRC:

```
[21:36:25] <SillySophist> so TIL: Everything about swap in LX is a lie?
[21:36:30] <pmooney> so yeah, we'll accept any input for swapiness
[21:37:39] <pmooney> sysctl 'vm.swappiness=gimmesomeswap'
[21:37:48] <pmooney> (returns 0)
[21:40:21] <SillySophist> interesting… I am wondering if some of these crazy slowdowns i’ve seen on CNs are related to apps not swapping
[21:41:05] <SillySophist> or maybe just trying to swap and taking too long
[21:41:25] <SillySophist> also this is amazing:
[21:41:36] <SillySophist> https://www.irccloud.com/pastebin/yj2EtKt8/
[21:42:47] <pmooney> iiiiinteresting...
[21:44:01] <SillySophist> https://github.com/joyent/illumos-joyent/blob/master/usr/src/uts/common/brand/lx/procfs/lx_prvnops.c#L3779-L3785
[21:44:03] <SillySophist> lol
[21:44:16] <SillySophist> that’s a great comment
[21:44:22] <SillySophist> pmooney: i know right!?
[21:44:38] <pmooney> pretty sure I know what's going wrong there
[21:44:49] <pmooney> You're slightly over the zone memcap
[21:44:55] <pmooney> so we're overflowing on the subtract
[21:45:12] <pmooney> Would you mind opening a bug on illumos-joyent?
[21:45:46] <pmooney> include the 'free' output and mention that the zone is completely at capacity from a memory standpoint (right?)
```

**Comment by chayeon** [comment]
Hi,
I have the same issue. For test I wrote following program. and run it in LX branded zones.
<pre>
#include &lt;stdlib.h&gt;
#include &lt;stdio.h&gt;
#include &lt;string.h&gt;

int
main(void) {
    unsigned int size = 10 * 1024 * 1024; //10M memory allocation
    char *allocMem;
    while (1) {
        allocMem = malloc(size);
        memset(allocMem, 'D', size);
        sleep(1);
        printf("DDDD\n");
    }
    return 0;
}
</pre>
And in another terminal I could see following result.
<pre>
top - 07:32:02 up  1:29,  0 users,  load average: 0.00, 0.00, 0.00
Tasks:   9 total,   1 running,   8 sleeping,   0 stopped,   0 zombie
%Cpu(s):  0.3 us,  2.8 sy,  0.0 ni, 96.9 id,  0.0 wa,  0.1 hi,  0.0 si,  0.0 st
KiB Mem:   1048576 total,   887768 used,   160808 free,        0 buffers
KiB Swap:  4194304 total,        0 used,  4194304 free.        0 cached Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
<b>40166 root      59   0 1502536 1.428g      0 S   0.9  143   0:01.18 a.out</b>
40705 root      59   0    7060   2228      0 S   0.1  0.2   0:00.00 sleep
89200 root      59   0    7164   2448      0 S   0.1  0.2   0:02.57 dash
40138 root      59   0   22532   3256      0 R   0.0  0.3   0:00.09 top
25143 root      59   0   20944   3668      0 S   0.0  0.3   0:00.01 bash
    1 root      59   0    7164   2480      0 S   0.0  0.2   0:00.11 dash
39893 root      59   0   20944   3664      0 S   0.0  0.3   0:00.01 bash
39987 root      59   0   20944   3656      0 S   0.0  0.3   0:00.01 bash
25591 root       1   0   20944   3668      0 S   0.0  0.3   0:00.00 bash
</pre>

<pre>
root@4cd54c5bc305:/# free
             total       used       free     shared    buffers     cached
<b>Mem:       1048576    1625476    -576900 </b>         0          0          0
-/+ buffers/cache:    1625476    -576900
Swap:      4194304          0    4194304
root@4cd54c5bc305:/# free <b>-m</b>
             total       used       free     shared    buffers     cached
<b>Mem:          1024       1587 18014398509481420</b>          0          0          0
-/+ buffers/cache:       1587 18014398509481420
Swap:         4096          0       4096
</pre>

The `free` says that the Swap is not used. Is that right?

Thank you.

**Comment by zadunn** [comment]
My understanding, gleaned in #smartos, is that basically we always lie to lx-zones about swap.  Illumos swap and linux swap are very different and attempting to map one to another is probably not a great idea.  

https://github.com/joyent/illumos-joyent/blob/master/usr/src/uts/common/brand/lx/procfs/lx_prvnops.c#L3638-L3643

https://github.com/joyent/illumos-joyent/blob/master/usr/src/uts/common/brand/lx/procfs/lx_prvnops.c#L4212-L4220


I think above is a better explanation.   I would still love to see this addressed since it blows up our metrics.

**Comment by jjelinek** [comment]
I opened OS-5836 for this issue.

**Comment by jjelinek** [comment]
Actually, we already have an open ticket OS-5741 which we file when this was first reported.

**Comment by jjelinek** [comment]
Fixed with commit eb3642a
